manager.py

```diff
@@ -126,280 +70,269 @@ def readme_item(item):
+    return protocols
 
-def try_connect(domain, url, connection):
-    try:
-        http = connection(domain, timeout=5)
-        http.request(
-            "GET",
-            url.path,
-            headers={
-                "User-Agent": "curl/8.0.1",
-            },
-        )
-        res = http.getresponse()
-        if res.status == 200:
-            return True
-    except Exception:
-        return False
-
-
-def try_protocols(mirror):
-    url = urlparse(mirror["url"])
-    domain = url.hostname
-    protocols = []
-    print(
-        'Accessing "{provider}" at "{domain}": ... '.format(domain=domain, **mirror),
-        end="",
-        flush=True,
-    )
 
-    try:
-        for family, _, _, _, sockaddr in socket.getaddrinfo(domain, 80):
-            ip = sockaddr[0]
-            ipa = ip_address(ip)
-            if ipa.is_global:
-                if type(ipa) is IPv4Address and "ipv4" not in protocols:
-                    protocols.append("ipv4")
-                if type(ipa) is IPv6Address and "ipv6" not in protocols:
-                    protocols.append("ipv6")
-        protocols.sort()
-    except socket.gaierror:
-        pass
-    else:
-        if try_connect(domain, url, HTTPConnection):
-            protocols.append("http")
-            url = url._replace(scheme="http")
-        if try_connect(domain, url, HTTPSConnection):
-            protocols.append("https")
-            url = url._replace(scheme="https")
-
-    print(", ".join(protocols))
-    mirror["protocols"] = protocols
-    mirror["url"] = urlunparse(url)
-
-
-def sub_protocols(args):
-    mirrors = []
-    with open(SOURCE_YAML, "r") as source:
-        try:
-            mirrors = ordered_load_yaml(source)
-        except yaml.YAMLError as e:
-            print(e)
-            sys.exit(1)
+def update_protocols(mirrors):
     for m in mirrors:
-        try_protocols(m)
-    with open(SOURCE_YAML, "w") as output:
-        print(
-            ordered_dump_yaml(
-                mirrors, encoding=None, allow_unicode=True, default_flow_style=False
-            ),
-            file=output,
+        url = urlparse(m["url"])
+        provider = m["provider"]
+        m["protocols"] = sorted(
+            get_http_connection(url, provider) | get_ip_version(url, provider)
         )
+        m["url"] = urlunparse(url)
 
 
-def sub_all(args):
-    sub_protocols(args)
-    sub_readme(args)
-    sub_mirrorlist(args)
-
+def sub_protocols(_):
+    mirrors = []
+    yaml = ruamel.yaml.YAML()
+    with open(SOURCE_YAML, encoding="utf-8") as source:
+        try:
+            mirrors = yaml.load(source)["archlinuxcn"]
+        except ruamel.yaml.YAMLError as error:
+            sys.exit(repr(error))
+    update_protocols(mirrors)
+    with open(SOURCE_YAML, "w", encoding="utf-8") as output:
+        yaml.explicit_start = True
+        yaml.indent(mapping=2, sequence=4, offset=2)
+        yaml.dump({"archlinuxcn": mirrors}, output)
```

Comment:
To preserve the original order when writing to yaml file, do not add the (typ="safe") argument.